src/openms/source/FORMAT/HANDLERS/MzMLSqliteHandler.cpp

```diff
@@ -0,0 +1,981 @@
+          else spec.getInstrumentSettings().setPolarity(IonSource::POSITIVE);
+        }
+        if (sqlite3_column_type(stmt, 15) != SQLITE_NULL && sqlite3_column_int(stmt, 15) != -1
+            && sqlite3_column_int(stmt, 15) < static_cast<int>(OpenMS::Precursor::SIZE_OF_ACTIVATIONMETHOD))
+        {
+          precursor.getActivationMethods().insert(static_cast<OpenMS::Precursor::ActivationMethod>(sqlite3_column_int(stmt, 15)));
+        }
+        if (sqlite3_column_type(stmt, 16) != SQLITE_NULL) precursor.setActivationEnergy(sqlite3_column_double(stmt, 16));
+
+        if (sqlite3_column_type(stmt, 6) != SQLITE_NULL) spec.getPrecursors().push_back(precursor);
+        if (sqlite3_column_type(stmt, 11) != SQLITE_NULL) spec.getProducts().push_back(product);
+        spectra.push_back(spec);
+
+        sqlite3_step( stmt );
+      }
+
+      // free memory
+      sqlite3_finalize(stmt);
+    }
+
+    void MzMLSqliteHandler::writeExperiment(const MSExperiment<> & exp)
+    {
+      writeChromatograms(exp.getChromatograms());
+      writeSpectra(exp.getSpectra());
+    }
+
+    void MzMLSqliteHandler::createTables()
+    {
+      sqlite3 *db;
+      char *zErrMsg = 0;
+      int rc;
+
+      // delete file if present
+      QFile file (filename_.toQString());
+      file.remove();
+
+      // Open database
+      rc = sqlite3_open(filename_.c_str(), &db);
+      if (rc)
+      {
+        throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, String("Can't open database: ") + sqlite3_errmsg(db));
+      }
+
+      // Create SQL structure
+      char const *create_sql =
+
+        // data table
+        //  - compression is one of 0 = no, 1 = zlib, 2 = np-linear, 3 = np-slof, 4 = np-pic, 5 = np-linear + zlib, 6 = np-slof + zlib, 7 = np-pic + zlib
+        //  - data_type is one of 0 = mz, 1 = int, 2 = rt
+        //  - data contains the raw (blob) data for a single data array
+        "CREATE TABLE DATA(" \
+        "SPECTRUM_ID INT," \
+        "CHROMATOGRAM_ID INT," \
+        "COMPRESSION INT," \
+        "DATA_TYPE INT," \
+        "DATA BLOB NOT NULL" \
+        ");" \
+
+        // spectrum table
+        "CREATE TABLE SPECTRUM(" \
+        "ID INT PRIMARY KEY NOT NULL," \
+        "MSLEVEL INT NULL," \
+        "RETENTION_TIME REAL NULL," \
+        "SCAN_POLARITY INT NULL," \
+        "NATIVE_ID TEXT NOT NULL" \
+        ");" \
+
+        // chromatogram table
+        "CREATE TABLE CHROMATOGRAM(" \
+        "ID INT PRIMARY KEY NOT NULL," \
+        "NATIVE_ID TEXT NOT NULL" \
+        ");" \
+
+        // product table
+        "CREATE TABLE PRODUCT(" \
+        "SPECTRUM_ID INT," \
+        "CHROMATOGRAM_ID INT," \
+        "CHARGE INT NULL," \
+        "ISOLATION_TARGET REAL NULL," \
+        "ISOLATION_LOWER REAL NULL," \
+        "ISOLATION_UPPER REAL NULL" \
+        ");" \
+
+        // precursor table
+        "CREATE TABLE PRECURSOR(" \
+        "SPECTRUM_ID INT," \
```

Comment:
well again design reasons. Also it makes it clearer if there is a mistake and you can never join a spectrum with a precursor of a chromatogram and vice versa. I think it prevents many "dumb" mistakes